zilliqa/src/pool.rs

```diff
@@ -0,0 +1,213 @@
+    }
+}
+
+impl Eq for ReadyItem {}
+
+impl PartialOrd for ReadyItem {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ReadyItem {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.gas_price.cmp(&other.gas_price)
+    }
+}
+
+impl From<&VerifiedTransaction> for ReadyItem {
+    fn from(txn: &VerifiedTransaction) -> Self {
+        ReadyItem {
+            gas_price: txn.tx.gas_price(),
+            from_addr: txn.signer,
+            nonce: txn.tx.nonce(),
+        }
+    }
+}
+
+impl TransactionPool {
+    /// Pop a *ready* transaction out of the pool, maximising the gas price.
+    ///
+    /// Ready means that the transaction has a nonce equal to the sender's current nonce or it has a nonce that is
+    /// consecutive with a previously returned transaction, from the same sender.
+    pub fn best_transaction(&mut self) -> Option<VerifiedTransaction> {
+        loop {
+            let Some(ReadyItem {
+                from_addr, nonce, ..
+            }) = self.ready.pop()
+            else {
+                return None;
+            };
+            let Some(transaction) = self.transactions.remove(&(from_addr, nonce)) else {
+                // A transaction might have been ready, but then we learnt that the sender's nonce increased, making it
+                // invalid. In this case, we will have removed the transaction from `transactions` in `update_nonce`.
+                // We loop until we find a transaction that hasn't been made invalid.
+                continue;
+            };
+            self.hash_to_index.remove(&transaction.hash);
+
+            if let Some(next_txn) = self.transactions.get(&(from_addr, nonce + 1)) {
+                self.ready.push(next_txn.into());
+            }
+
+            return Some(transaction);
+        }
+    }
+
+    pub fn insert_transaction(&mut self, transaction: VerifiedTransaction, account_nonce: u64) {
+        if transaction.tx.nonce() < account_nonce {
+            // This transaction is permanently invalid, so there is nothing to do.
+            return;
+        }
+
+        if transaction.tx.nonce() == account_nonce {
+            self.ready.push((&transaction).into());
+        }
+
+        self.hash_to_index.insert(
+            transaction.hash,
+            (transaction.signer, transaction.tx.nonce()),
+        );
+        self.transactions
+            .insert((transaction.signer, transaction.tx.nonce()), transaction);
+    }
+
+    pub fn get_transaction(&self, hash: Hash) -> Option<&VerifiedTransaction> {
+        let Some((addr, nonce)) = self.hash_to_index.get(&hash) else {
+            return None;
+        };
+
+        self.transactions.get(&(*addr, *nonce))
+    }
+
+    pub fn update_nonce(&mut self, address: Address, nonce: u64) {
+        // Remove a transaction from this sender with the previous nonce, if it exists.
+        self.transactions.remove(&(address, nonce - 1));
```

Comment:
I might just have to go with `fn increment_nonce()`. While it would be nicer to take a `&txn` and verify that it has only incremented the account's nonce by 1, I don't think we should rely on the fact that every node's mempool will know about all transactions before they are proposed. Admittedly, I think this is guaranteed for the current implemention, but it wouldn't be a very trivial change for a fork to not broadcast transactions and keep a private mempool (in fact we've suggested doing this with a configuration flag for frontrunning prevention).

Does that justification make sense to you?

Reply:
Oh actually that's nonsense, because I need to know some nonce to actually update the pool with. Your suggestion makes sense to me upon a second reading. I'll probably keep the name as `fn update_nonce` but just take a `&VerifiedTransaction` instead. I'll also add a comment saying that this needs to be called for all transactions, otherwise permanently invalidated transactions might be left in the pool.